File: demoTransport.py
'''
Ji-Eun Byun (ji-eun.byun@glasgow.ac.uk)
Created: 27 Mar 2023

A small, hypothetical bridge system
'''
import numpy as np
import networkx as nx
from scipy.stats import lognorm
import matplotlib
import pdb


matplotlib.use("TKAgg")
import matplotlib.pyplot as plt

from BNS_JT import cpm, variable
from Trans.trans import get_arcs_length, get_all_paths_and_times

## Data
# Network
node_coords = {1: (-2, 3),
               2: (-2, -3),
               3: (2, -2),
               4: (1, 1),
               5: (0, 0)}

arcs = {1: [1, 2],
	2: [1, 5],
	3: [2, 5],
	4: [3, 4],
	5: [3, 5],
	6: [4, 5]}

# Fragility curves -- From HAZUS-EQ model (roads are regarded as disconnected when being extensively or completely damaged)

frag = {'major': {'med': 60.0, 'std': 0.7},
        'urban' : {'med': 24.0, 'std': 0.7},
        'bridge': {'med': 1.1, 'std': 3.9},
        }

arcs_type = {1: 'major',
             2: 'major',
             3: 'major',
             4: 'urban',
             5: 'bridge',
             6: 'bridge'}

arcs_avg_kmh = {1: 40,
                2: 40,
                3: 40,
                4: 30,
                5: 30,
                6: 20}

ODs = [(5, 1), (5, 2), (5, 3), (5, 4)]

# For the moment, we assume that ground motions are observed. Later, hazard nodes will be added.
GM_obs = {1: 30.0,
          2: 20.0,
          3: 10.0,
          4: 2.0,
          5: 0.9,
          6: 0.6}

# Arcs' states index
arc_surv = 1
arc_fail = 2
arc_either = 3

arc_lens_km = get_arcs_length(arcs, node_coords)

#arcTimes_h = arcLens_km ./ arcs_Vavg_kmh
arc_times_h = {k: v/arcs_avg_kmh[k] for k, v in arc_lens_km.items()}

G = nx.Graph()
for k, x in arcs.items():
    G.add_edge(x[0], x[1], time=arc_times_h[k], label=k)

for k, v in node_coords.items():
    G.add_node(k, pos=v)

path_time = get_all_paths_and_times(ODs, G, key='time')

# plot graph
pos = nx.get_node_attributes(G, 'pos')
edge_labels = nx.get_edge_attributes(G, 'label')

fig = plt.figure()
ax = fig.add_subplot()
nx.draw(G, pos, with_labels=True, ax=ax)
nx.draw_networkx_edge_labels(G, pos, edge_labels=edge_labels, ax=ax)
fig.savefig("graph.png", dpi=200)

# Arcs (components)
cpms_arc = {}
vars_arc = {}
for k in arcs.keys():
    _type = arcs_type[k]
    prob = lognorm.cdf(GM_obs[k], frag[_type]['std'], scale=frag[_type]['med'])
    C = np.array([[arc_surv, arc_fail]]).T
    p = np.array([1-prob, prob])
    cpms_arc[k] = cpm.Cpm(variables = [k],
                          no_child = 1,
                          C = C,
                          p = p)

    B = np.array([[1, 0], [0, 1], [1, 1]])
    vars_arc[k] = variable.Variable(B=B, value=['Surv', 'Fail'])

# Travel times (systems)
c7 = np.array([
[1,3,1,3,3,3,3],
[2,1,2,1,3,3,3],
[3,1,2,2,3,3,3],
[3,2,2,3,3,3,3]])

c8 = np.array([
[1,3,3,1,3,3,3],
[2,1,1,2,3,3,3],
[3,1,2,2,3,3,3],
[3,2,3,2,3,3,3]])

c9 = np.array([
[1,3,3,3,3,1,3],
[2,3,3,3,1,2,1],
[3,3,3,3,1,2,2],
[3,3,3,3,2,2,3]])

c10 = np.array([
[1,3,3,3,3,3,1],
[2,3,3,3,1,1,2],
[3,3,3,3,1,2,2],
[3,3,3,3,2,3,2]])

cpms_arc[7] = cpm.Cpm(variables= [7, 1, 2, 3, 4, 5, 6],
                       no_child = 1,
                       C = c7,
                       p = [1, 1, 1, 1],
                       )

cpms_arc[8] = cpm.Cpm(variables= [8, 1, 2, 3, 4, 5, 6],
                       no_child = 1,
                       C = c8,
                       p = [1, 1, 1, 1],
                       )

cpms_arc[9] = cpm.Cpm(variables= [9, 1, 2, 3, 4, 5, 6],
                       no_child = 1,
                       C = c9,
                       p = [1, 1, 1, 1],
                       )

cpms_arc[10] = cpm.Cpm(variables= [10, 1, 2, 3, 4, 5, 6],
                       no_child = 1,
                       C = c10,
                       p = [1, 1, 1, 1],
                       )

B_ = np.array([[1, 0, 0], [0, 1, 0], [0, 0, 1]])
vars_arc[7] = variable.Variable(B=B_,
        value=[0.0901, 0.2401, np.inf])

vars_arc[8] = variable.Variable(B=B_,
        value=[0.0901, 0.2401, np.inf])

vars_arc[9] = variable.Variable(B=B_,
        value=[0.0943, 0.1761, np.inf])

vars_arc[10] = variable.Variable(B=B_,
        value=[0.0707, 0.1997, np.inf])

## Inference - by variable elimination (would not work for large-scale systems)
# Probability of delay and disconnection
cpms_arc_cp = cpms_arc.values()

for i in arcs.keys():

    isinscope_tf = cpm.isinscope([i], cpms_arc_cp)
    cpm_sel = [y for x, y in zip(isinscope_tf, cpms_arc_cp) if x]
    cpm_mult, vars_arc = cpm.prod_cpms(cpm_sel, vars_arc)
    cpm_mult = cpm_mult.sum([i])

    cpms_arc_cp = [y for x, y in zip(isinscope_tf, cpms_arc_cp) if x == False]
    cpms_arc_cp.insert(0, cpm_mult)

# Retrieve example results
ODs_prob_disconn = np.zeros(len(ODs))
ODs_prob_delay = np.zeros(len(ODs))

var_OD = [7, 8, 9, 10]
for j, idx in enumerate(var_OD):

    # Prob. of disconnection
    disconn_state = vars_arc[idx].B.shape[0]
    [cpm_ve], vars_arc = cpm.condition(cpms_arc_cp,
                           cnd_vars=[idx],
                           cnd_states=[disconn_state],
                           var=vars_arc)
    ODs_prob_disconn[j] = cpm_ve.p.sum(axis=0)

    # Prob. of delay
    var_loc = np.where(cpms_arc_cp[0].variables == idx)[0]
    rows_to_keep = np.where(cpms_arc_cp[0].C[:, var_loc] > 1)[0]
    cpm_ve = cpms_arc_cp[0].get_subset(rows_to_keep)
    ODs_prob_delay[j] = cpm_ve.p.sum(axis=0)

# City 1 and 2 experienced a disruption in getting resources, City 3 was okay and 4 is unknown. Probability of damage of roads?
# A composite state needs be created for City 1 and City 2
for idx in var_OD:
    vars_arc[idx].B = np.vstack([vars_arc[idx].B, [0, 1, 1]])

# # Add observation nodes
var_OD_obs = []
for j, (idx, od) in enumerate(zip(var_OD, ODs), start=11):

    #iPaths = arcPaths[iOdInd]
    #iTimes = arcPaths_time[iOdInd]
    cpms_arc[j]= cpm.Cpm(variables=[j, idx],
                           no_child=1,
                           C= np.array([[1, 1], [2, 2], [2, 3]]),
                           p= [1, 1, 1])
    vars_arc[j] = variable.Variable(B=np.eye(2, dtype=int), value=['No disruption', 'Disruption'])

    var_OD_obs.append(j)

cpm_ve, vars_arc = cpm.condition(cpms_arc,
                                 cnd_vars=[11, 12, 13],
                                 cnd_states=[2, 2, 1],
                                 var=vars_arc)

#pdb.set_trace()

Mcond_mult, vars_arc = cpm.prod_cpms(cpm_ve, vars_arc)
Mcond_mult_sum = Mcond_mult.sum(var_OD + var_OD_obs)

arcs_prob_damage = np.zeros(len(arcs))
for j, i in enumerate(arcs.keys()):

    iM = Mcond_mult_sum.sum([i], 0)
    [iM_fail], vars_arc = cpm.condition(iM, [i], [arc_fail], vars_arc)
    fail_prob = iM_fail.p / iM.p.sum(axis=0)
    if fail_prob.any():
        arcs_prob_damage[j] = fail_prob

# plot
plt.figure()

# set width of bars
barWidth = 0.25

# Set position of bar on X axis
r1 = np.arange(len(ODs))
r2 = [x + barWidth for x in r1]

# Make the plot
plt.bar(r1, ODs_prob_disconn, color='#7f6d5f', width=barWidth, edgecolor='white', label='Disconnection')
plt.bar(r2, ODs_prob_delay, color='#557f2d', width=barWidth, edgecolor='white', label='Delay')

# Add xticks on the middle of the group bars
plt.xlabel('OD pair', fontweight='bold')
plt.xticks([r + 0.5*barWidth for r in range(len(ODs))], [f'{x}' for x in ODs])

# Create legend & Show graphic
plt.legend()
plt.savefig('./figure1.png', dpi=200)
plt.close()

plt.figure()

r1 = list(arcs.keys())
r2 = [x + barWidth for x in r1]

plt.bar(r1, 1-arcs_prob_damage, color='#7f6d5f', width=barWidth, edgecolor='white', label='Survival')
plt.bar(r2, arcs_prob_damage, color='#557f2d', width=barWidth, edgecolor='white', label='Failure')

plt.xlabel( 'Arc' )
plt.ylabel( 'Probability' )
plt.legend()
plt.savefig('./figure2.png', dpi=200)
plt.close()

File: demoTransport.m
%{
Ji-Eun Byun (ji-eun.byun@glasgow.ac.uk)
Created: 27 Mar 2023

A small, hypothetical bridge system
%}

%% Data
% Network
nodeCoords_km = [-2 3; -2 -3; 2 -2; 1 1; 0 0];
arcs = [1 2; 1 5; 2 5; 3 4; 3 5; 4 5];
major = 1; urban = 2; bridge = 3;
arcsType = [major; major; major; urban; bridge; bridge];
arcs_Vavg_kmh = [40; 40; 40; 30; 30; 20];
ODs = [5 1; 5 2; 5 3; 5 4];
arc_surv = 1; arc_fail=2; arc_either=3; % Arcs' states index

arcLens_km = funTrans.getArcsLength( arcs, nodeCoords_km );
arcTimes_h = arcLens_km ./ arcs_Vavg_kmh;
G = graph(arcs(:,1), arcs(:,2), arcTimes_h);
[arcPaths, arcPaths_time] = funTrans.getAllPathsAndTimes( ODs, G, arcTimes_h );


% GM
GM_obs = [30 20 10 2 0.9 0.6]; % For the moment, we assume that ground motions are observed. Later, hazard nodes will be added.

% Fragility curves -- From HAZUS-EQ model (roads are regarded as disconnected when being extensively or completely damaged)
IMs = [60 24 1.1];
IMs_std = [0.7 0.7 3.9];


%% BN set up
varInd = 0;
M = Cpm;
vars = Variable;

% Arcs (components): P(X_i | GM = GM_ob ), i = 1 .. N (= nArc)
nArc = size(arcs,1);
var_arcs = zeros(1,nArc);
for iArcInd = 1:nArc
    varInd = varInd+1;

    iRoadType = arcsType(iArcInd);
    iFrag_mean = IMs(iRoadType); iFrag_std = IMs_std(iRoadType);
    iGM_obs = GM_obs(iArcInd);
    iFailProb = funTrans.normcdf( 1/iFrag_std * log( iGM_obs/iFrag_mean ));

    M(varInd,1) = Cpm( varInd, 1, [arc_surv; arc_fail], [1-iFailProb; iFailProb] );
    vars(varInd,1) = Variable( [eye(2); 1 1], {'Surv'; 'Fail'} );
    var_arcs(iArcInd) = varInd;
end

% Travel times (systems): P( OD_j | X_1 ... X_N ), j = 1 .. M (= nOD)
nOD = size(ODs,1);
var_OD = zeros(1,nOD);
for iOdInd = 1:nOD
    varInd = varInd+1;

    iPaths = arcPaths{iOdInd};
    iTimes = arcPaths_time{iOdInd};

    [iC, iP, iVals] = funTrans.getCPMOfTravelTime( iPaths, iTimes, nArc, arc_surv, arc_fail, arc_either );
    M(varInd,1) = Cpm( [varInd var_arcs], 1, iC, iP );
    vars(varInd,1) = Variable( eye(length(iVals)), iVals );
    var_OD(iOdInd) = varInd;
end


%% Inference - by variable elimination (would not work for large-scale systems)
% Probability of delay and disconnection
varElimOrder = var_arcs;
M_VE = M; % Becomes P(OD_1, ..., OD_M) since X_1, ..., X_N are eliminated
for iVarInd_ = 1:length(varElimOrder)
    iVarInd = varElimOrder(iVarInd_);
    iIsVarInScope = isXinScope( iVarInd, M_VE );
    
    [iMMult, vars] = multCPMs(M_VE(iIsVarInScope), vars);
    iMMult = sum( iMMult, iVarInd );

    M_VE(iIsVarInScope) = [];
    M_VE = [iMMult; M_VE];
end


% Retrieve example results
ODs_prob_disconn = zeros(1,nOD); % P( OD_j = 3 ), j = 1, ..., M, where State 3 indicates disconnection
ODs_prob_delay = zeros(1,nOD); % P( (OD_j = 2) U (OD_j = 2) ), j = 1, ..., M, where State 2 indicates the use of the second shortest path (or equivalently, P( (OD_j = 1)^c ), where State 1 indicates the use of the shortest path)
for iODInd = 1:nOD
    iVarInd = var_OD(iODInd);

    % Prob. of disconnection
    iDisconnState = size(vars(iVarInd).B,1);
    iM_VE = condition( M_VE, iVarInd, iDisconnState, vars );

    iDisconnProb = sum( iM_VE.p );
    ODs_prob_disconn(iODInd) = iDisconnProb;

    % Prob. of delay
    iVarLocInC = find(M_VE.variables == iVarInd);
    iRowsIndToKeep = find( M_VE.C(:, iVarLocInC)> 1 );
    iM_VE = getCpmSubset( M_VE, iRowsIndToKeep );

    iDelayProb = sum( iM_VE.p );
    ODs_prob_delay(iODInd) = iDelayProb;
end

figure;
bar( [ODs_prob_disconn(:) ODs_prob_delay(:)] )

grid on
xlabel( 'OD pair' )
ylabel( 'Probability' )
legend( {'Disconnection' 'Delay'}, 'location', 'northwest' )


% City 1 and 2 experienced a disruption in getting resources, City 3 was okay and 4 is unknown. Probability of damage of roads?
% % A composite state needs be created for City 1 and City 2
for iODInd = 1:nOD
    vars( var_OD(iODInd) ).B = [vars( var_OD(iODInd) ).B; 0 1 1];
end

% % Add observation nodes P( O_j | OD_j ), j = 1, ..., M
Mobs = M;
var_ODobs = zeros(1,nOD);
for iOdInd = 1:nOD
    varInd = varInd+1;

    iPaths = arcPaths{iOdInd};
    iTimes = arcPaths_time{iOdInd};

    Mobs(varInd,1) = Cpm( [varInd var_OD(iOdInd)], 1, [1 1; 2 2; 2 3], [1; 1; 1] );
    vars(varInd,1) = Variable( eye(2), {'No disruption'; 'Disruption'} );

    var_ODobs(iOdInd) = varInd;
end

[Mcond, vars] = condition(Mobs,[var_ODobs(1) var_ODobs(2) var_ODobs(3)], [2 2 1], vars);
[Mcond_mult, vars] = multCPMs(Mcond, vars);
Mcond_mult_sum = sum( Mcond_mult, [var_OD var_ODobs] );

arcs_prob_damage = zeros(1,nArc); % P( X_i = 2 | OD_1 = 2, OD_2 = 2, OD_3 = 1 ), i = 1, ..., N
for iArcInd = 1:nArc
    iVarInd = var_arcs(iArcInd);

    iM = sum( Mcond_mult_sum, iVarInd, 0 );
    [iM_fail, vars] = condition( iM, iVarInd, arc_fail, vars );
    
    iFailProb = iM_fail.p / sum(iM.p);
    if ~isempty(iFailProb)
        arcs_prob_damage(iArcInd) = iFailProb;
    else
        arcs_prob_damage(iArcInd) = 0;
    end
end


figure;
bar( [1-arcs_prob_damage(:) arcs_prob_damage(:)] )

grid on
xlabel( 'Arc' )
ylabel( 'Probability' )
legend( {'Survive' 'Fail'}, 'location', 'northwest' )


%% Repeat inferences again using new functions -- the results must be the same.
% Probability of delay and disconnection
[M_VE2, vars] = VE( M, varElimOrder, vars ); % "M_VE2" same as "M_VE"

% Retrieve example results
ODs_prob_disconn2 = zeros(1,nOD); % "ODs_prob_disconn2" same as "ODs_prob_disconn"
ODs_prob_delay2 = zeros(1,nOD); % "ODs_prob_delay2" same as "ODs_prob_delay"
for iODInd = 1:nOD
    iVarInd = var_OD(iODInd);

    % Prob. of disconnection
    iDisconnState = find( arrayfun(@(x) x==100, vars(iVarInd).v) ); % the state of disconnection is assigned an arbitrarily large number 100
    iDisconnProb = getProb( M_VE2, iVarInd, iDisconnState, vars );
    ODs_prob_disconn2(iODInd) = iDisconnProb;

    % Prob. of delay
    iDelayProb = getProb( M_VE2, iVarInd, 1, vars, 0 ); % Any state greater than 1 means delay.
    ODs_prob_delay2(iODInd) = iDelayProb;
end

% Check if the results are the same
disp( ['Are the disconnected probabilities the same?: ' num2str( isequal( ODs_prob_disconn2, ODs_prob_disconn ) )] )
disp( ['Are the delay probabilities the same?: ' num2str( isequal( ODs_prob_delay2, ODs_prob_delay ) )] )